src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -0,0 +1,430 @@
+
+  while (freeSpace < size) { // TODO: Think about parallel reads and writes; can this turn into an infinite loop? 
+    CacheBlock* victim = find_victim(dpp, y);
+
+    if (victim == nullptr) {
+      ldpp_dout(dpp, 10) << "LFUDAPolicy::" << __func__ << "(): Could not retrieve victim block." << dendl;
+      delete victim;
+      return -1;
+    }
+
+    std::string key = victim->cacheObj.bucketName + "_" + victim->cacheObj.objName + "_" + std::to_string(victim->blockID) + "_" + std::to_string(victim->size);
+    auto it = entries_map.find(key);
+    if (it == entries_map.end()) {
+      delete victim;
+      return -1;
+    }
+
+    int avgWeight = get_local_weight_sum(y) / entries_map.size();
+    if (avgWeight < 0) {
+      delete victim;
+      return -1;
+    }
+
+    if (victim->hostsList.size() == 1 && victim->hostsList[0] == dir->cct->_conf->rgw_local_cache_address) { /* Last copy */
+      if (victim->globalWeight) {
+	const std::lock_guard l(lfuda_lock);
+	it->second->localWeight += victim->globalWeight;
+        (*it->second->handle)->localWeight = it->second->localWeight;
+	entries_heap.increase(it->second->handle);
+
+	if (cacheDriver->set_attr(dpp, key, "user.rgw.localWeight", std::to_string(it->second->localWeight), y) < 0) 
+	  return -1;
+
+	victim->globalWeight = 0;
+	if (dir->update_field(victim, "globalWeight", std::to_string(victim->globalWeight), y) < 0) {
+	  delete victim;
+	  return -1;
+        }
+      }
+
+      if (it->second->localWeight > avgWeight) {
+	// TODO: push victim block to remote cache
+	// add remote cache host to host list
+      }
+    }
+
+    victim->globalWeight += it->second->localWeight;
+    if (dir->update_field(victim, "globalWeight", std::to_string(victim->globalWeight), y) < 0) {
+      delete victim;
+      return -1;
+    }
+
+    if (dir->remove_host(victim, dir->cct->_conf->rgw_local_cache_address, y) < 0) {
+      delete victim;
+      return -1;
+    }
+
+    delete victim;
+
+    if (cacheDriver->del(dpp, key, y) < 0) 
+      return -1;
+
+    ldpp_dout(dpp, 10) << "LFUDAPolicy::" << __func__ << "(): Block " << key << " has been evicted." << dendl;
+
+    int weight = (avgWeight * entries_map.size()) - it->second->localWeight;
+    if (set_local_weight_sum((weight > 0) ? weight : 0, y) < 0)
+      return -1;
+
+    int age = get_age(y);
+    age = std::max(it->second->localWeight, age);
+    if (set_age(age, y) < 0)
+      return -1;
+
+    erase(dpp, key, y);
+    freeSpace = cacheDriver->get_free_space(dpp);
+  }
+  
+  return 0;
+}
+
+void LFUDAPolicy::update(const DoutPrefixProvider* dpp, std::string& key, uint64_t offset, uint64_t len, std::string version, optional_yield y)
+{
+  using handle_type = boost::heap::fibonacci_heap<LFUDAEntry*, boost::heap::compare<EntryComparator<LFUDAEntry>>>::handle_type;
+
+  int age = get_age(y); 
+  int localWeight = age;
+  auto entry = find_entry(key);
+  if (entry != nullptr) { 
+    entry->localWeight += age;
```

Comment:
@samarahu : shouldn't localWeight increase by 1 on every access, before adding cache age to it? else localWeight of all blocks will always remain 0, the initial value of cache age is 0, and while eviction, the new cache age is based on the local weight of the evicted block and the cache age, both of which are 0, so the new cache age will also be zero.

Reply:
@samarahu : did you address the above comment

Reply:
I updated this so that `entry->localWeight` is not altered before the `erase` call. Instead, a copy is stored in the `localWeight` variable declared on line 331 and has the `age` added to it. It is later used to set the updated `localWeight` for the new entry.